ufl/algorithms/signature.py

```diff
@@ -93,21 +93,22 @@ def compute_terminal_hashdata(expressions, renumbering):
     return terminal_hashdata
 
 
-def compute_expression_hashdata(expression, terminal_hashdata):
-    # The hashdata computed here can be interpreted as prefix operator
-    # notation, i.e. we store the equivalent of '+ * a b * c d' for
-    # the expression (a*b)+(c*d)
-    expression_hashdata = []
-    for expr in pre_traversal(expression):
+def compute_expression_hashdata(expression, terminal_hashdata) -> bytes:
+    cache = {}
+
+    for expr in unique_post_traversal(expression):
+        # Uniquely traverse tree and hash each node
+        # E.g. (a + b*c) is hashed as hash([+, hash(a), hash([*, hash(b), hash(c)])])
+        # Traversal uses post pattern, so children hashes are cached
         if expr._ufl_is_terminal_:
-            data = terminal_hashdata[expr]
+            data = [terminal_hashdata[expr]]
         else:
-            data = expr._ufl_typecode_  # TODO: Use expr._ufl_signature_data_()? More extensible, but more overhead.
-        expression_hashdata.append(data)
-    # Oneliner: TODO: Benchmark, maybe use a generator?
-    # expression_hashdata = [(terminal_hashdata[expr] if expr._ufl_is_terminal_ else expr._ufl_typecode_)
-    #                       for expr in pre_traversal(expression)]
-    return expression_hashdata
+            data = [expr._ufl_typecode_]
+
+            for op in expr.ufl_operands:
+                data += [cache[op]]
+        cache[expr] = hashlib.sha512(str(data).encode("utf-8")).digest()
```

Comment:
Is it necessary to hash every operand, or is it sufficient to strify? I guess if the cost of this approach is basically zero then it doesn't really matter too much.

Reply:
The cost of hashing is not significant. It is necessary to hash the tuple or operands, since their stringification would explode the memory (my crazy UFL form got always `Killed`). So I need to compress the already visited expression subtree with hashlib.

Reply:
Thanks, makes sense.